### PR TITLE
On 276 create new submission events endpoint

### PR DESCRIPTION
**Pull request by davidgouge**

### CODE REVIEW

File: src/api/SFA.DAS.Provider.Events.Api.IntegrationTests/Setup/DatabaseSetup.cs [hunk 1]
```diff
@@ -36,6 +39,9 @@ public DatabaseSetup(PopulateTables populate)
         {
             Debug.WriteLine("Populating submission events table, this could take a while");
 
+            await PopulateEventsForByUln(1000000160, 2);
```
Comment (MichaelYoung1981) [inline src/api/SFA.DAS.Provider.Events.Api.IntegrationTests/Setup/DatabaseSetup.cs]: Does this databaset setup method get called for each test?  If it does, we don't want to insert the ULN events each time if most tests don't need them.  If it's a one off call for each assembly then it's fine.

File: src/api/SFA.DAS.Provider.Events.Api.IntegrationTests/Setup/DatabaseSetup.cs [hunk 2]
```diff
@@ -71,7 +95,7 @@ public DatabaseSetup(PopulateTables populate)
 
         private async Task ReadSubmissionEvents()
         {
-            const string submissionEventsSql = "SELECT * FROM [Submissions].[SubmissionEvents]";
+            const string submissionEventsSql = "SELECT * FROM [Submissions].[SubmissionEvents] WHERE Id > 3";
```
Comment (MichaelYoung1981) [inline src/api/SFA.DAS.Provider.Events.Api.IntegrationTests/Setup/DatabaseSetup.cs]: Not sure using the id to filter is a good idea.  Could we use uln = null instead?

File: src/api/SFA.DAS.Provider.Events.Api/Controllers/SubmissionsController.cs [hunk 3]
```diff
@@ -59,5 +60,39 @@ public SubmissionsController(IMediator mediator, ILogger logger)
...
+                var queryResponse = await _mediator.SendAsync(new GetLatestLearnerEventByStandardQueryRequest
+                    {
+                        SinceEventId = sinceEventId,
+                        Uln = uln
+                    })
+                    .ConfigureAwait(false);
+                
+                if (!queryResponse.IsValid)
+                {
+                    throw queryResponse.Exception;
```
Comment (MichaelYoung1981) [inline src/api/SFA.DAS.Provider.Events.Api/Controllers/SubmissionsController.cs]: I know this is the existing pattern in the solution but it's awful (and loses the stack trace!).  If there is an exception it should be thrown from the query rather than added to the response message and then thrown.

File: src/api/SFA.DAS.Provider.Events.Api/Web.config [hunk 4]
```diff
@@ -18,9 +18,9 @@
   <appSettings>
     <add key="LogLevel" value="Debug" />
     <add key="EnvironmentName" value="LOCAL" />
-    <add key="LoggingConnectionString" value="server=.;database=DasProviderEventsApi;trusted_connection=true;" />
-    <add key="MonthEndConnectionString" value="server=.;database=DasProviderEventsApi;trusted_connection=true;" />
-    <add key="EventsConnectionString" value="server=.;database=DasProviderEventsApi;trusted_connection=true;" />
+    <add key="LoggingConnectionString" value="server=.;database=ProviderEventsAT;trusted_connection=true;" />
```
Comment (MichaelYoung1981) [inline src/api/SFA.DAS.Provider.Events.Api/Web.config]: You will want to revert these changes otherwise it'll stop the API running for everything else (unless they rename their databases).

File: src/api/SFA.DAS.Provider.Events.Application.UnitTests/Submissions/GetLatestLearnerEventByStandardQuery/GetLatestLearnerEventByStandardQueryHandler/WhenHandling.cs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            // Assert
+            Assert.IsNotNull(actual);
+            Assert.IsFalse(actual.IsValid);
+            Assert.IsInstanceOf<Exception>(actual.Exception);
+        }
+
+        [Test]
+        public void ThenItShouldReturnValidResults()
+        {
+            _submissionEventsRepository.Setup(r => r.GetLatestLearnerEventByStandard(It.IsAny<long>(), It.IsAny<long>()))
```
Comment (MichaelYoung1981) [inline src/api/SFA.DAS.Provider.Events.Application.UnitTests/Submissions/GetLatestLearnerEventByStandardQuery/GetLatestLearnerEventByStandardQueryHandler/WhenHandling.cs]: Can It.IsAny be changed to the actual expected parameters?  It'll make the test a lot tighter.

File: src/api/SFA.DAS.Provider.Events.Application/Submissions/GetLatestLearnerEventByStandardQuery/GetLatestLearnerEventByStandardQueryHandler.cs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+                return new GetLatestLearnerEventByStandardQueryResponse
+                {
+                    IsValid = true,
+                    Result = entities
+                };
+            }
+            catch (Exception ex)
+            {
+                return new GetLatestLearnerEventByStandardQueryResponse
```
Comment (MichaelYoung1981) [inline src/api/SFA.DAS.Provider.Events.Application/Submissions/GetLatestLearnerEventByStandardQuery/GetLatestLearnerEventByStandardQueryHandler.cs]: As per the comment on the controller, exceptions shouldn't be handled in this way, it can just bubble up.

File: src/api/SFA.DAS.Provider.Events.Api.IntegrationTests/Setup/DatabaseSetup.cs (hunk 1 above)
Comment (davidgouge) [reply src/api/SFA.DAS.Provider.Events.Api.IntegrationTests/Setup/DatabaseSetup.cs]: This is only called once at the start.

**Comment by davidgouge** [comment]
Made changes as per @MichaelYoung1981's suggestions.

File: src/api/SFA.DAS.Provider.Events.Api/Controllers/SubmissionsController.cs [hunk 7]
```diff
@@ -59,5 +60,34 @@ public SubmissionsController(IMediator mediator, ILogger logger)
                 return InternalServerError();
             }
         }
+
+        [Route("api/v2/submissions", Name = "GetLatestLearnerEventByStandard")]
```
Comment (Pauljgraham) [inline src/api/SFA.DAS.Provider.Events.Api/Controllers/SubmissionsController.cs]: There doesn't appear to be any unit tests against this controller action.
Also of concern is the class 'SubmissionEventEntity' is being exposed directly as a result. There is no API Type and hence no API Client Code to call this endpoint. (This may be a problem with some existing endpoints, but I haven't checked)
I'm not clear on how ValidationException could be thrown. I was unable to test the endpoint locally, so I could not test is this is thrown. This failure to test locally, is partly down to the CloudService Role being mis-configured (pre-existing issue).

File: src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsSubmissionEventsRepository.cs [hunk 8]
```diff
@@ -72,6 +75,22 @@ public DcfsSubmissionEventsRepository()
                 .ConfigureAwait(false);
         }
 
+        public async Task<IEnumerable<SubmissionEventEntity>> GetLatestLearnerEventByStandard(long eventId, long uln)
```
Comment (Pauljgraham) [inline src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsSubmissionEventsRepository.cs]: Just a question, if we are only expecting to return a single record. Why are we returning this a list which will only have one record or none?

File: src/api/SFA.DAS.Provider.Events.Api.IntegrationTests/Setup/DatabaseSetup.cs [hunk 9]
```diff
@@ -53,6 +60,24 @@ public DatabaseSetup(PopulateTables populate)
...
         }
 
+        private async Task PopulateEventsForByUln(long uln, int count)
+        {
+            TestData.SubmissionEventsForUln = new Fixture().CreateMany<ItSubmissionEvent>(count).ToList();
+            foreach (var @event in TestData.SubmissionEventsForUln)
+            {
+                @event.Uln = uln;
+                // these fields are just date's in the db
+                @event.ActualStartDate = @event.ActualStartDate.Value.Date;
```
[the remark this answers is not in the document]
Comment (davidgouge) [reply src/api/SFA.DAS.Provider.Events.Api.IntegrationTests/Setup/DatabaseSetup.cs]: This section was duplicated from the existing method PopulateSubmissionEvents().  I didn't want to assume anything, so left those lines as they were.

File: src/api/SFA.DAS.Provider.Events.Api/Controllers/SubmissionsController.cs (hunk 7 above)
Comment (davidgouge) [reply src/api/SFA.DAS.Provider.Events.Api/Controllers/SubmissionsController.cs]: Looking into changing the type being returned and adding unit tests now.

File: src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsSubmissionEventsRepository.cs (hunk 8 above)
Comment (davidgouge) [reply src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsSubmissionEventsRepository.cs]: Good question, looks like the IEnumerable is a hangover to when we ere going to return many results.  Fixing up.

Comment (davidgouge) [reply src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsSubmissionEventsRepository.cs]: Double checked and the query can indeed return more than one result if the learner has more than one standard code in the Submission events.  The rownumber = 1 is just limiting the results to one per standard code.

File: src/api/SFA.DAS.Provider.Events.Api/Controllers/SubmissionsController.cs (hunk 7 above)
Comment (davidgouge) [reply src/api/SFA.DAS.Provider.Events.Api/Controllers/SubmissionsController.cs]: Latest push to this branch adds unit tests for the Controller Action, changes the return type to be the correct one, updates the API Client to include the new Action (and added client tests) and Validation is now handled properly.

Comment (Pauljgraham) [reply src/api/SFA.DAS.Provider.Events.Api/Controllers/SubmissionsController.cs]: Looks good to me, just one question. How do you run the solution locally as my fails when trying to run just the API?

File: src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsSubmissionEventsRepository.cs (hunk 8 above)
Comment (Pauljgraham) [reply src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsSubmissionEventsRepository.cs]: OK by me

**Comment by satya-sanikommu** [comment]
Review completed, merging this into 'BaseBranch_DPP-1086_1030_1378' to test 4 features together

**Comment by satya-sanikommu** [comment]
Ignore the previous comment, in a wrong PR

**Comment by satya-sanikommu** [comment]
@carlbm PR updated post your suggestions, please review this when you get some time

File: src/api/SFA.DAS.Provider.Events.Api.UnitTests/Controllers/SubmissionsController/WhenGettingLatestLearnerEventsByStandard.cs [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+        [Test]
+        public async Task ThenItShouldReturnCorrectListOfEvents()
+        {
+            // Act
+            var actual = ((OkNegotiatedContentResult<List<SubmissionEvent>>)await _controller.GetLatestLearnerEventByStandard(1111111111)).Content;
+
+            // Assert
+            actual.Should().NotBeNull();
+            actual.Count.Should().Be(2);
```
Comment (carlbm) [inline src/api/SFA.DAS.Provider.Events.Api.UnitTests/Controllers/SubmissionsController/WhenGettingLatestLearnerEventsByStandard.cs]: Why the mix of nunit assert and fluentassertions?

File: src/api/SFA.DAS.Provider.Events.Api/Controllers/LearnersController.cs [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+        private ILogger _logger;
+
+        public LearnersController(IMediator mediator, ILogger logger)
+        {
+            _mediator = mediator;
+            _logger = logger;
+        }
+
+
+        [Route("api/v2/learners", Name = "GetLatestLearnerEventByStandard")]
```
Comment (carlbm) [inline src/api/SFA.DAS.Provider.Events.Api/Controllers/LearnersController.cs]: Why V2?

File: src/api/SFA.DAS.Provider.Events.Api/Controllers/LearnersController.cs [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        [HttpGet]
+        public async Task<IHttpActionResult> GetLatestLearnerEventByStandard(long uln, long sinceEventId = 0)
```
Comment (carlbm) [inline src/api/SFA.DAS.Provider.Events.Api/Controllers/LearnersController.cs]: The name of the method makes it sound like we are doing it by standard, or that there should be a standard code as a parameter. Can you rename it to either ...ForStandards or leave the standard part out?

File: src/api/SFA.DAS.Provider.Events.Application/Submissions/GetLatestLearnerEventByStandardQuery/GetLatestLearnerEventByStandardQueryHandler.cs [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+using System.Threading.Tasks;
+using MediatR;
+using SFA.DAS.Provider.Events.Api.Types;
+using SFA.DAS.Provider.Events.Application.Mapping;
+using SFA.DAS.Provider.Events.Application.Repositories;
+using SFA.DAS.Provider.Events.Application.Validation;
+
+namespace SFA.DAS.Provider.Events.Application.Submissions.GetLatestLearnerEventByStandardQuery
+{
+    public class GetLatestLearnerEventByStandardQueryHandler :
```
Comment (carlbm) [inline src/api/SFA.DAS.Provider.Events.Application/Submissions/GetLatestLearnerEventByStandardQuery/GetLatestLearnerEventByStandardQueryHandler.cs]: See comment on the API endpoint about the name

File: src/api/SFA.DAS.Provider.Events.Application/Submissions/GetLatestLearnerEventByStandardQuery/GetLatestLearnerEventByStandardQueryRequestValidator.cs [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace SFA.DAS.Provider.Events.Application.Submissions.GetLatestLearnerEventByStandardQuery
+{
+    public class GetLatestLearnerEventByStandardQueryRequestValidator : IValidator<GetLatestLearnerEventByStandardQueryRequest>
+    {
+        public async Task<ValidationResult> Validate(GetLatestLearnerEventByStandardQueryRequest item)
+        {
+            var validationErrors = new[]
+            {
+                item.Uln < 1000000000 || item.Uln > 9999999999 ? "Uln must be between 1111111111 and 9999999999": null
```
Comment (carlbm) [inline src/api/SFA.DAS.Provider.Events.Application/Submissions/GetLatestLearnerEventByStandardQuery/GetLatestLearnerEventByStandardQueryRequestValidator.cs]: I thought there was more to a valid ULN? At least a checksum?

**Comment by carlbm** [comment]
Also, please update the readme files to document the new endpoint

File: src/api/SFA.DAS.Provider.Events.Application/Submissions/GetLatestLearnerEventByStandardQuery/GetLatestLearnerEventByStandardQueryRequestValidator.cs (hunk 14 above)
Comment (carlbm) [reply src/api/SFA.DAS.Provider.Events.Application/Submissions/GetLatestLearnerEventByStandardQuery/GetLatestLearnerEventByStandardQueryRequestValidator.cs]: Sat says that there is a checksum. Can you find out what it is and include it please?

File: src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsSubmissionEventsRepository.cs [hunk 15]
```diff
@@ -79,7 +79,7 @@ public DcfsSubmissionEventsRepository()
         {
             var eventIdFilterClause = eventId > 0 ? $"se.Id > {eventId} AND " : "";
 
-            var command = $@"SELECT* FROM(
+            var command = $@"SELECT *, CommitmentId AS ApprenticeshipId FROM(
                 SELECT ROW_NUMBER() OVER (PARTITION BY se.StandardCode ORDER BY se.Id DESC) rownumber, se.*
                 FROM Submissions.SubmissionEvents se
             WHERE {eventIdFilterClause} se.ULN = {uln}) subEvents
```
Comment (carlbm) [inline src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsSubmissionEventsRepository.cs]: Could you remove the string interpolation and instead use a sql parameter. Dapper supports it and it's generally used in the rest of the codebase. It's importand here as this is a user provided field that we're using to access the database

Comment (davidgouge) [reply src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsSubmissionEventsRepository.cs]: "used in the rest of the codebase"

image

Comment (davidgouge) [reply src/api/SFA.DAS.Provider.Events.Infrastructure/Data/DcfsSubmissionEventsRepository.cs]: Done.

**Comment by satya-sanikommu** [comment]
Review and Testing completed. Ready to merge
